packages/core/src/agent/getOutboundMessageContext.ts

```diff
@@ -0,0 +1,247 @@
+import type { AgentMessage } from './AgentMessage'
+import type { AgentContext } from './context'
+import type { ConnectionRecord, Routing } from '../modules/connections'
+import type { BaseRecord } from '../storage/BaseRecord'
+
+import { Key } from '../crypto'
+import { ServiceDecorator } from '../decorators/service/ServiceDecorator'
+import { AriesFrameworkError } from '../error'
+import { OutOfBandService, OutOfBandRole, OutOfBandRepository } from '../modules/oob'
+import { OutOfBandRecordMetadataKeys } from '../modules/oob/repository/outOfBandRecordMetadataTypes'
+import { RoutingService } from '../modules/routing'
+import { DidCommMessageRepository, DidCommMessageRole } from '../storage'
+import { uuid } from '../utils/uuid'
+
+import { OutboundMessageContext } from './models'
+
+/**
+ * Maybe these methods should be moved to a service, but that would require
+ * extra injection in the sender functions, and I'm not 100% sure what would
+ * be the best place to host these.
+ */
+
+/**
+ * Get the outbound message context for a message. Will use the connection record if available,
+ * and otherwise try to create a connectionless message context.
+ */
+export async function getOutboundMessageContext(
+  agentContext: AgentContext,
+  {
+    message,
+    connectionRecord,
+    associatedRecord,
+    lastReceivedMessage,
+    lastSentMessage,
+  }: {
+    connectionRecord?: ConnectionRecord
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    associatedRecord?: BaseRecord<any, any, any>
```

Comment:
Could we have these `any`s in the `BaseRecord` as defaults for the generic? Usage would be a bit cleaner and we only have to disable `any` at that point.

Reply:
We could also take a generic for `getOutboundMessageContext` as when we call this function we always know the associatedRecord type, right?

Reply:
having any on the record messes with the types when you extend the `BaseRecord` (so then you always have to provide types when extending). 

We can use the generic, but then you still need to do `BaseRecord<any, any, any>` on the generic